pallets/parachain-staking/src/migrations.rs

```diff
@@ -37,6 +37,104 @@ use frame_support::{
+			Delegations<T::AccountId, BalanceOf<T>>,
+			Twox64Concat,
+		>(pallet_prefix, top_delegations_prefix)
+		.collect();
+		let migrated_candidates_top_count: Weight = stored_top_delegations
+			.len()
+			.try_into()
+			.expect("There are between 0 and 2**64 mappings stored.");
+		let stored_bottom_delegations: Vec<_> = storage_key_iter::<
+			T::AccountId,
+			Delegations<T::AccountId, BalanceOf<T>>,
+			Twox64Concat,
+		>(pallet_prefix, bottom_delegations_prefix)
+		.collect();
+		let migrated_candidates_bottom_count: Weight = stored_bottom_delegations
+			.len()
+			.try_into()
+			.expect("There are between 0 and 2**64 mappings stored.");
+		fn fix_delegations<T: Config>(
+			delegations: Delegations<T::AccountId, BalanceOf<T>>,
+		) -> Delegations<T::AccountId, BalanceOf<T>> {
+			let correct_total = delegations
+				.delegations
+				.iter()
+				.fold(BalanceOf::<T>::zero(), |acc, b| acc + b.amount);
+			log::info!(
+				target: "PatchIncorrectDelegationSums",
+				"Correcting total from {:?} to {:?}",
+				delegations.total, correct_total
+			);
+			Delegations {
+				delegations: delegations.delegations,
+				total: correct_total,
+			}
+		}
+		for (account, old_top_delegations) in stored_top_delegations {
+			let new_top_delegations = fix_delegations::<T>(old_top_delegations);
+			let mut candidate_info = <CandidateInfo<T>>::get(&account)
+				.expect("TopDelegations exists => CandidateInfo exists");
+			candidate_info.total_counted = candidate_info.bond + new_top_delegations.total;
+			if candidate_info.is_active() {
+				Pallet::<T>::update_active(account.clone(), candidate_info.total_counted);
+			}
+			<CandidateInfo<T>>::insert(&account, candidate_info);
+			<TopDelegations<T>>::insert(&account, new_top_delegations);
+		}
+		for (account, old_bottom_delegations) in stored_bottom_delegations {
+			let new_bottom_delegations = fix_delegations::<T>(old_bottom_delegations);
+			<BottomDelegations<T>>::insert(&account, new_bottom_delegations);
+		}
+		let weight = T::DbWeight::get();
+		let top = migrated_candidates_top_count.saturating_mul(3 * weight.write + 3 * weight.read);
+		let bottom = migrated_candidates_bottom_count.saturating_mul(weight.write + weight.read);
+		// 20% max block weight as margin for error
+		top + bottom + 100_000_000_000
+	}
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		// get delegation count for all candidates
+		for (account, state) in <CandidateInfo<T>>::iter() {
+			let total_delegation_count = state;
+			Self::set_temp_storage(
+				state.delegation_count,
+				&format!("Candidate{:?}DelegationCount", account)[..],
+			);
+		}
+		Ok(())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade() -> Result<(), &'static str> {
+		// ensure delegation count was unchanged by migration
+		for (account, state) in <CandidateInfo<T>>::iter() {
+			let expected_count: u32 =
+				Self::get_temp_storage(&format!("Candidate{:?}DelegationCount", account)[..])
```

Comment:
We probably should do a more thorough check here. We are expeting to total_counted for top to equal the sum of all delegations for the top candidate right? I think we should check that

Reply:
The comprehensive migration test is in the pallet unit test.

> We are expeting to total_counted for top to equal the sum of all delegations for the top candidate right? I think we should check that

We are expecting total_counted for the collator to equal sum of all top delegations + collator self bond. I'll write this, but I don't expect to ever run it. I use the unit test for my testing and I never run the `try-runtime` checks.

Reply:
Yeah I saw the unit test, but I thought better have it here too. Thanks!